### PR TITLE
Added support for Loading Filtered Policies from DB

### PR DESCRIPTION
**Pull request by vinod827**
Hi,

I have coded a new method called loadFilteredPolicy() to load filtered policies from a given entity repo based on filtered object passed to it.
We are using TypeORM Adapter for Casbin in our project and this is an urgent requirement for us to have this missing functionality so I have added in it and raised PR.

Kindly assist to review my PR and approve it.

Thank you,
Vinod Kumar

### CODE REVIEW

**Comment by hsluoyz** [comment]
@nodece @ghaiklor @GopherJ please review.

**Comment by nodece** [comment]
Please fix the CI error.

**Comment by vinod827** [comment]
> Please fix the CI error.

sorry, is this comment (for CI error) for me to check?....If yes then I tried to analyse as much possible from my side however I did not find anything relevant with respect to my code...I believe this is generic issue of CI/CD pipeline....

**Comment by nodece** [comment]
@vinod827 You implement `FilteredAdapter`.

File: test/adapter.test.ts [hunk 1]
```diff
@@ -125,6 +125,10 @@ test('TestAdapter', async () => {
             ['bob', 'data2', 'write'],
             ['data2_admin', 'data2', 'read'],
             ['data2_admin', 'data2', 'write']]);
+
+        // Load Filtered Policy
+        await a.loadFilteredPolicy(e.getModel(), { ptype : 'p', v0 : 'alice' });
```
Comment (nodece) [inline test/adapter.test.ts:130]: You should call `e.loadFilteredPolicy`.

```js
// @ts-ignore - :( node-casbin provides an error type. I will fix it after that.
await e.loadFilteredPolicy({ ptype : 'p', v0 : 'alice' });
```

**Comment by nodece** [comment]
##  How to implement filtered adapter?

We need to abide by the following rules. 

### loadPolicy

When the method is done, set `filtered` is `false`.

### loadFilteredPolicy

When the method is done, set `filtered` is `true`.

> Note: Casbin can not update policy when enable filtered model.

File: test/adapter.test.ts (hunk 1 above)
Comment (vinod827) [reply test/adapter.test.ts:130]: okay

**Comment by vinod827** [comment]
> ## How to implement filtered adapter?
> We need to abide by the following rules.
> 
> ### loadPolicy
> When the method is done, set `filtered` is `false`.
> 
> ### loadFilteredPolicy
> When the method is done, set `filtered` is `true`.
> 
> > Note: Casbin can not update policy when enable filtered model.

Is there any example code to set this filtered flag to true?

**Comment by vinod827** [comment]
Hi All,

Please check the build. Only 1 is failing due to some issue. 
https://travis-ci.org/github/node-casbin/typeorm-adapter/builds/691784303

Kindly assist to review the PR as well.
Thanks

**Comment by nodece** [comment]
Hi @vinod827 , example FilteredAdapter:
````ts
class TestAdapter implements FilteredAdapter {
    private filtered = false;

    isFiltered(): boolean {
        return this.filtered;
    }

    async loadPolicy(model: Model): Promise<void> {
        // do something - load policy from db
        
        this.filtered = false;
        return undefined;
    }

    async addPolicy(sec: string, ptype: string, rule: string[]): Promise<void> {
        if (this.filtered) {
            throw new Error("un-support removePolicy in filter model")
        }

        // do something
    }

    async loadFilteredPolicy(model: Model, filter: any): Promise<void> {
        // do something - load policy from db

        this.filtered = true;
    }

    async removeFilteredPolicy(sec: string, ptype: string, fieldIndex: number, ...fieldValues: string[]): Promise<void> {
        if (this.filtered) {
            throw new Error("un-support removePolicy in filter model")
        }

        // do something
    }

    async removePolicy(sec: string, ptype: string, rule: string[]): Promise<void> {
        if (this.filtered) {
            throw new Error("un-support removePolicy in filter model")
        }

        // do something
    }

    async savePolicy(model: Model): Promise<boolean> {
        if (this.filtered) {
            throw new Error("un-support removePolicy in filter model")
        }

        // do something
        return true;
    }
}

```

**Comment by nodece** [comment]
@vinod827 You can remove node6 in CI. mysql2 doesn't support node6.

**Comment by hsluoyz** [comment]
Please do not append commits to fix your errors. can you rewrite previous commits?

image

**Comment by vinod827** [comment]
> Please do not append commits to fix your errors. can you rewrite previous commits?
> 
> image

Sure, I will merge into a larger commit.

File: test/adapter.test.ts [hunk 2]
```diff
@@ -125,6 +125,13 @@ test('TestAdapter', async () => {
             ['bob', 'data2', 'write'],
             ['data2_admin', 'data2', 'read'],
             ['data2_admin', 'data2', 'write']]);
+
+        // Load Filtered Policy
+        /*await a.loadFilteredPolicy(e.getModel(), {
```
Comment (nodece) [inline test/adapter.test.ts:130]: @vinod827 Remove the comments. Run it on test.

File: README.md [hunk 3]
```diff
@@ -70,6 +70,46 @@ async function myFunction() {
...
+    const e = await newEnforcer('examples/rbac_model.conf', a);
+
+    // Load the filtered policy from DB.
+    await e.loadFilteredPolicy({
+        'ptype': 'p',
+        'v0': 'alice'
+    });
+
+    // Check the permission.
+    e.enforce('alice', 'data1', 'read');
```
Comment (nodece) [inline README.md:103]: miss `await`

File: src/adapter.ts [hunk 4]
```diff
@@ -23,14 +23,19 @@ type CasbinRuleConstructor = new (...args: any[]) => GenericCasbinRule;
 /**
  * TypeORMAdapter represents the TypeORM adapter for policy storage.
  */
-export default class TypeORMAdapter implements Adapter {
+export default class TypeORMAdapter implements Adapter, FilteredAdapter {
```
Comment (nodece) [inline src/adapter.ts:26]: good: `export default class TypeORMAdapter implements FilteredAdapter`

**Comment by nodece** [comment]
@vinod827 Why remove the `loadFilteredPolicy`? You should add some tests for `loadFilteredPolicy`.

File: README.md (hunk 3 above)
Comment (vinod827) [reply README.md:103]: okay, will add

File: test/adapter.test.ts (hunk 2 above)
Comment (vinod827) [reply test/adapter.test.ts]: okay

**Comment by vinod827** [comment]
> @vinod827 Why remove the `loadFilteredPolicy`? You should add some tests for `loadFilteredPolicy`.

okay, will add test, thanks

File: src/adapter.ts (hunk 4 above)
Comment (vinod827) [reply src/adapter.ts]: okay

**Comment by vinod827** [comment]
Hi Team,

I have incorporated all the review comments as suggested and pushed the changes again. This time only build failed due to some other CI error:-
https://travis-ci.org/github/node-casbin/typeorm-adapter/jobs/692144298

I would really appreciate if you can please have a look and approve my PR after review as this functionality is really important and urgent for us.

Thank you,
Regards,
Vinod Kumar

**Comment by nodece** [comment]
LGTM, thank  @vinod827 for your contribution! 

 Could you rewrite your commits? It's looks chaos :(

After that, I will merge it into the master.

**Comment by hsluoyz** [comment]
@nodece why the CI is still failing?

**Comment by nodece** [comment]
@hsluoyz Due mysql2 doesn't support node@6: https://travis-ci.org/github/node-casbin/typeorm-adapter/jobs/692144298#L217

**Comment by hsluoyz** [comment]
@vinod827 @nodece then we should remove node 6 from Travis CI config file.

**Comment by hsluoyz** [comment]
@ghaiklor sounds good.

@nodece can you help remove Node 6 and bump to a new release?

**Comment by nodece** [comment]
@vinod827 @hsluoyz `v1.2.0` has been released.

@ghaiklor `mysql2` doesn't support Node.js v6 but it as `devDependency.  Maybe publish it as minor release is better.

**Comment by ghaiklor** [comment]
Hm, I thought, `mysql` is the dependency for the library to operate. Makes sense then to publish as a minor release, if it is just a dev dependency for testing.

**Pull request merged**